backend-rust/src/indexer.rs

```diff
@@ -335,96 +335,112 @@ impl Indexer for BlockPreProcessor {
+                    match tokenomics_info.response {
+                        RewardsOverview::V0 {
+                            ..
+                        } => {
+                            compute_total_stake_capital(
+                                &mut client6,
+                                BlockIdentifier::AbsoluteHeight(fbi.height),
+                            )
+                            .await
+                        }
+                        RewardsOverview::V1 {
+                            total_staked_capital,
+                            ..
+                        } => Ok(total_staked_capital),
+                    }
+                };
+                let get_block_certificates = async move {
+                    // Fetching the block certificates prior to P6 results in a NotFound gRPC error,
+                    // so we produce the empty type of certificates instead.
+                    if block_info.response.protocol_version < ProtocolVersion::P6 {
+                        Ok(BlockCertificates {
+                            quorum_certificate:       None,
+                            timeout_certificate:      None,
+                            epoch_finalization_entry: None,
+                        })
+                    } else {
+                        let response = client7.get_block_certificates(fbi.height).await?;
+                        Ok(response.response)
+                    }
+                };
 
-            let start_fetching = Instant::now();
-            let (
-                block_info,
-                chain_parameters,
-                events,
-                items,
-                tokenomics_info,
-                special_events,
-                certificates,
-            ) = try_join!(
-                client1.get_block_info(fbi.height),
-                client2.get_block_chain_parameters(fbi.height),
-                get_events,
-                get_items,
-                client.get_tokenomics_info(fbi.height),
-                get_special_items,
-                client6.get_block_certificates(fbi.height)
-            )?;
-            let total_staked_capital = match tokenomics_info.response {
-                RewardsOverview::V0 {
-                    ..
-                } => {
-                    compute_total_stake_capital(
-                        &mut client,
-                        BlockIdentifier::AbsoluteHeight(fbi.height),
-                    )
-                    .await?
-                }
-                RewardsOverview::V1 {
-                    total_staked_capital,
-                    ..
-                } => total_staked_capital,
-            };
-            let node_response_time = start_fetching.elapsed();
-            self.node_response_time.get_or_create(label).observe(node_response_time.as_secs_f64());
+                let (total_staked_capital, certificates) =
+                    try_join!(get_total_staked_capital, get_block_certificates)?;
```

Comment:
These should be done within the block where it is needed so we only wait on that instead of waiting on all of the async blocks to finish before we can proceed with this one

Reply:
Yes, you are right.
Will improve this